### PR TITLE
CC-166 | Some improvements, new DevkitUtils class

### PR DESCRIPTION
**Pull request by ssaucedo**
@CaetanoR  I just reorganized the code and covered some possible bugs. Let me know if it makes sense for you.
@gastonortiz please check NodeIterable.class
:shipit:

### CODE REVIEW

File: src/main/java/org/mule/tools/devkit/sonar/utils/DevkitUtils.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    public static String getLatestDevKitVersion() {
+        String latestVersion = null;
+        try (InputStream xml = new URL("https://repository.mulesoft.org/nexus/content/repositories/releases/org/mule/tools/devkit/mule-devkit-parent/maven-metadata.xml").openStream()) {
+            Document doc = DocumentBuilderFactory.newInstance()
+                    .newDocumentBuilder()
+                    .parse(xml);
+            doc.getDocumentElement()
+                    .normalize();
+            latestVersion = getMayorVersion(filter(transform(new NodeIterable(doc.getElementsByTagName("version")), getVersion), and(notNull(), isValidVersion)));
```
Comment (urbanusjam) [inline src/main/java/org/mule/tools/devkit/sonar/utils/DevkitUtils.java]: Mayor --> Major

File: src/main/java/org/mule/tools/devkit/sonar/utils/DevkitUtils.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        } catch (ParserConfigurationException | IOException | SAXException e) {
+            logger.error("Unable to retrieve the XML for DevKit metadata from Nexus repository. The rule 'DevKitLatestVersionCheck' won't be executed at this point", e);
+        }
+        return latestVersion;
+    }
+
+    private static String EMPTY = "";
```
Comment (urbanusjam) [inline src/main/java/org/mule/tools/devkit/sonar/utils/DevkitUtils.java]: Is this variable needed? Why not use StringUtils.EMPTY directly?

File: src/main/java/org/mule/tools/devkit/sonar/utils/DevkitUtils.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    private static final Logger logger = LoggerFactory.getLogger(PomUtils.class);
+
+    public static Predicate<String> isValidVersion = new Predicate<String>() {
+
+        @Override
+        public boolean apply(@Nullable String version) {
+            return !isRevision(version)
+                    && !version.startsWith("4");
```
Comment (urbanusjam) [inline src/main/java/org/mule/tools/devkit/sonar/utils/DevkitUtils.java]: `startsWith` can throw NPE. Add a nullcheck for "version".

File: src/main/java/org/mule/tools/devkit/sonar/utils/NodeIterable.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            Integer cursor = 0;
+
+            @Override
+            public boolean hasNext() {
+                return cursor != nodeList.getLength() - 1;
+            }
+
+            @Override
+            public Node next() {
+                cursor = new Integer(cursor + 1);
```
Comment (urbanusjam) [inline src/main/java/org/mule/tools/devkit/sonar/utils/NodeIterable.java]: cursor = cursor + 1

File: src/main/java/org/mule/tools/devkit/sonar/utils/NodeIterable.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+            }
+        };
+    }
+
+    public static Function<Node, String> getVersion = new Function<Node, String>() {
+
+        @Override
+        public String apply(@Nullable Node node) {
+            if (node.getNodeType() == ELEMENT_NODE) {
```
Comment (urbanusjam) [inline src/main/java/org/mule/tools/devkit/sonar/utils/NodeIterable.java]: Add a null-check for "node".

File: src/main/java/org/mule/tools/devkit/sonar/utils/NodeIterable.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                return node.getFirstChild()
+                        .getTextContent();
+            }
+            return null;
```
Comment (urbanusjam) [inline src/main/java/org/mule/tools/devkit/sonar/utils/NodeIterable.java]: Can be converted into a single-line statement:

```java
return (node != null && node.getNodeType() == ELEMENT_NODE) ? node.getFirstChild().getTextContent() : null;
```

File: src/main/java/org/mule/tools/devkit/sonar/utils/NodeIterable.java (hunk 4 above)
Comment (gastonortiz) [inline src/main/java/org/mule/tools/devkit/sonar/utils/NodeIterable.java]: cursor++;
:)

**Pull request merged**